sheetsync/sheetsync/main.py

```diff
@@ -0,0 +1,266 @@
+				except ValueError:
+					value = None
+			row_dict[column] = value
+		return row_dict
+
+	def sync_row(self, worksheet, row_index, row, event):
+		"""Take a row dict and an Event from the database (or None if id not found)
+		and take whatever action is required to sync them, ie. writing to the database or sheet."""
+
+		if event is None:
+			# No event currently in DB, if any field is non-empty, then create it.
+			# Otherwise ignore it.
+			if not any(row[col] for col in self.input_columns):
+				return
+			logging.info("Inserting new event {}".format(row['id']))
+			# Insertion conflict just means that another sheet sync beat us to the insert.
+			# We can ignore it.
+			insert_cols = ['id'] + self.input_columns
+			built_query = sql.SQL("""
+				INSERT INTO events ({})
+				VALUES ({})
+				ON CONFLICT DO NOTHING
+			""").format(
+				sql.SQL(", ").join(sql.Identifier(col) for col in insert_cols),
+				sql.SQL(", ").join(sql.Placeholder(col) for col in insert_cols),
+			)
+			query(self.conn, built_query, **row)
+			return
+
+		# Update database with any changed inputs
+		changed = [col for col in self.input_columns if row[col] != getattr(event, col)]
+		if changed:
+			logging.info("Updating event {} with new value(s) for {}".format(
+				row['id'], ', '.join(changed)
+			))
+			built_query = sql.SQL("""
+				UPDATE events
+				SET {}
+				WHERE id = %(id)s
+			""").format(sql.SQL(", ").join(
+				sql.SQL("{} = {}").format(
+					sql.Identifier(col), sql.Placeholder(col)
+				) for col in changed
+			))
+			query(self.conn, built_query, **row)
+
+		# Update sheet with any changed outputs
+		format_output = lambda v: '' if v is None else v # cast nulls to empty string
+		changed = [col for col in self.output_columns if row[col] != format_output(getattr(event, col))]
+		if changed:
+			logging.info("Updating sheet row {} with new value(s) for {}".format(
+				row['id'], ', '.join(changed)
+			))
+			for col in changed:
+				self.sheets.write_value(
+					self.sheet_id, worksheet,
+					row_index, self.column_map[col],
+					format_output(getattr(event, col)),
+				)
+
+		# Set edit link if marked for editing and start/end set.
+		# This prevents accidents / clicking the wrong row and provides
+		# feedback that sheet sync is still working.
+		# Also clear it if it shouldn't be set.
+		edit_link = self.edit_url.format(row['id']) if row['marked_for_edit'] == '[+] Marked' else ''
+		if row['edit_link'] != edit_link:
+			logging.info("Updating sheet row {} with edit link {}".format(row['id'], edit_link))
+			self.sheets.write_value(
+				self.sheet_id, worksheet,
+				row_index, self.column_map['edit_link'],
+				edit_link,
+			)
+
+
+@argh.arg('worksheet-names', nargs='+', help="The names of the individual worksheets within the sheet to operate on.")
+def main(dbconnect, sheets_creds_file, edit_url, bustime_start, sheet_id, worksheet_names, metrics_port=8004, backdoor_port=0, allocate_ids=False):
```

Comment:
The thrimshim is already on port 8004 if we want to keep the different components on different ports.